src/instructlab/sdg/datamixing.py

```diff
@@ -615,10 +683,31 @@ def collect(
             output_file_leaf_skills = (
                 f"node_datasets_{self.date_suffix}/{leaf_node_path}_p10.jsonl"
             )
+            # HACK(osilkin): `knowledge_upsample_amount` is currently used when the generated knowledge data
+            #                 is orders of magnitude smaller (approx. < 3%) than the skills dataset.
+            #                 It is used to upsample that dataset so that the model doesn't forget it in training.
+            #
+            #                 This work around is currently hacky as we lack insight into the size of both datasets
+            #                 when we generate this data, and it may vary across different scenarios
+            sampling_size: int | float = 1.0
+            if self._precomputed_skills_length:
+                knowledge_to_skills_ratio = (
+                    len(skills_phase_data) / self._precomputed_skills_length
+                )
+                if knowledge_to_skills_ratio < MIN_UPSAMPLE_THRESHOLD:
+                    sampling_size = int(self._precomputed_skills_length * 0.03)
+
+                    print(
+                        "\033[93m"
+                        + f"Knowledge detected to be less than {MIN_UPSAMPLE_THRESHOLD*100:.2f}% of skills ({knowledge_to_skills_ratio:.2f}%), upsampling to: {sampling_size}"
+                        + "\033[0m"
+                    )
```

Comment:
Please swap this `print(...)` out for a log message, since we use `logger` elsewhere.

Reply:
Will do!

Reply:
Done

Reply:
Thanks!